jni/src/Foreign/JNI.hs

```diff
@@ -470,52 +543,82 @@ getStaticMethodID
   -> JNI.String -- ^ JNI signature
   -> IO JMethodID
 getStaticMethodID cls methodname sig = withJNIEnv $ \env ->
+    withJ cls $ \clsp ->
     throwIfException env $
     JNI.withString methodname $ \methodnamep ->
     JNI.withString sig $ \sigp ->
     [CU.exp| jmethodID {
       (*$(JNIEnv *env))->GetStaticMethodID($(JNIEnv *env),
-                                           $(jclass cls),
+                                           $(jclass clsp),
                                            $(char *methodnamep),
                                            $(char *sigp)) } |]
 
 getObjectClass :: Coercible o (J ty) => o -> IO JClass
 getObjectClass (coerce -> upcast -> obj) = withJNIEnv $ \env ->
-    [CU.exp| jclass {
+    withJ obj $ \objp ->
+    newJ =<< [CU.exp| jclass {
       (*$(JNIEnv *env))->GetObjectClass($(JNIEnv *env),
-                                        $(jobject obj)) } |]
+                                        $(jobject objp)) } |]
 
+-- | Creates a global reference to the object referred to by
+-- the given reference.
+--
+-- Arranges for a finalizer to call 'deleteGlobalRef' when the
+-- global reference is no longer reachable on the Haskell side.
 newGlobalRef :: Coercible o (J ty) => o -> IO o
 newGlobalRef (coerce -> upcast -> obj) = withJNIEnv $ \env ->
-    fmap coerce $
-    [CU.exp| jobject {
-      (*$(JNIEnv *env))->NewGlobalRef($(JNIEnv *env),
-                                      $(jobject obj)) } |]
+    fmap coerce $ withJ obj $ newGlobalJ env
 
 deleteGlobalRef :: Coercible o (J ty) => o -> IO ()
-deleteGlobalRef (coerce -> upcast -> obj) = withJNIEnv $ \env ->
-    [CU.block| void {
-      (*$(JNIEnv *env))->DeleteGlobalRef($(JNIEnv *env),
-                                         $(jobject obj)); } |]
+deleteGlobalRef (coerce -> J p) = finalizeForeignPtr p
+
+-- | Gets a local reference to the given java value.
+getLocalRefPtr :: Coercible o (J ty) => o -> IO (Ptr (J ty))
```

Comment:
I'd like that, but it is not possible to use it in `jvm-streaming`'s `posStream` which requires returning a pointer. Unless you do something like this:
```Haskell
j <- reflect x
withJ j return
```
Which is unsafe in general, but here it works because we count on `reflect` yielding a local reference with no finalizers to delete it.